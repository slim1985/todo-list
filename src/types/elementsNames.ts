export enum ElementsNames {
    STATUS_SELECT = 'status-select',
    TITLE_TEXT_AREA = 'title-textarea',
    DESCRIPTION_TEXT_AREA = 'description-textarea',
}
