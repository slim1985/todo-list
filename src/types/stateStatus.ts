export enum StateStatus {
    LOADING = 'loading',
    IDLE = 'idle',
    FAILED = 'failed',
}
