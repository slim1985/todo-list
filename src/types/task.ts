export interface Task {
    id: string;
    title: string;
    description: string;
    state: TaskStates;
}

export enum TaskStates {
    new = 'new',
    active = 'active',
    completed = 'completed',
}
