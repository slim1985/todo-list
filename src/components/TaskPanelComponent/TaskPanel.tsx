import { User } from '../../types/user';

export interface TaskPanelProps {
    user: User | null;
    signOut: () => void;
    clearTaskList: () => void;
    openTask: (taskId: string | null) => void;
}

export function TaskPanel({
    user,
    signOut,
    clearTaskList,
    openTask,
}: TaskPanelProps): JSX.Element {
    function signOutUser(): void {
        signOut();
        clearTaskList();
    }

    return (
        <div className="flex w-screen justify-start sticky top-0 space-x-3 bg-white solid border-gray-400 border-b-2">
            {user && (
                <>
                    <div className="flex w-2/4 justify-start items-center float-left">
                        <div className="ml-2">
                            <p className="font-semibold">{user?.displayName}</p>
                            <p className="text-sm">{user?.email}</p>
                            <button
                                onClick={() => signOutUser()}
                                className="mt-2"
                            >
                                Sign Out
                            </button>
                        </div>
                    </div>
                    <div className="flex w-2/4 justify-end">
                        <button
                            onClick={() => openTask(null)}
                            className="size-20 my-2 mr-2 p-1 bg-gray-300 rounded-md"
                        >
                            Create
                        </button>
                    </div>
                </>
            )}
        </div>
    );
}
