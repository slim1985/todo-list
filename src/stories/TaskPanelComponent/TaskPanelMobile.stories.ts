import type { Meta, StoryObj } from '@storybook/react';
import { TaskPanel } from '../../components/TaskPanelComponent/TaskPanel';

type Story = StoryObj<typeof meta>;

const meta: Meta<typeof TaskPanel> = {
    component: TaskPanel,
    parameters: {
        viewport: {
            defaultViewport: 'mobile1',
        },
    },
};

export default meta;

export const TaskPanelEmpty: Story = {
    name: 'Empty',
    args: {
        user: null,
    },
};

export const TaskPanelFilled: Story = {
    name: 'Filled',
    args: {
        user: {
            id: '1',
            displayName: 'John Doe',
            email: 'john.doe@gmail.com',
        },
    },
};
