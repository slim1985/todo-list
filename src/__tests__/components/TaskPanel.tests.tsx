import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { TaskPanel } from '../../components/TaskPanelComponent/TaskPanel';
import { User } from '../../types/user';

const user: User = {
    id: '1',
    displayName: 'Test User',
    email: 'user@user.com',
};

const signOut = jest.fn();
const clearTaskList = jest.fn();
const openTask = jest.fn();

describe('TaskPanel', () => {
    it('TaskPanel renders correctly with user', () => {
        // Arrange.
        // Act.
        const { asFragment } = render(
            <TaskPanel
                user={user}
                signOut={signOut}
                clearTaskList={clearTaskList}
                openTask={openTask}
            />,
        );

        // Assert.
        expect(asFragment()).toMatchSnapshot();
    });

    it('TaskPanel renders correctly without user', () => {
        // Arrange.
        // Act.
        const { asFragment } = render(
            <TaskPanel
                user={null}
                signOut={signOut}
                clearTaskList={clearTaskList}
                openTask={openTask}
            />,
        );

        // Assert.
        expect(asFragment()).toMatchSnapshot();
    });

    it('TaskPanel user data renders correctly', () => {
        // Arrange.
        // Act.
        render(
            <TaskPanel
                user={user}
                signOut={signOut}
                clearTaskList={clearTaskList}
                openTask={openTask}
            />,
        );

        // Assert.
        expect(screen.getByText(user.displayName)).toBeInTheDocument();
        expect(screen.getByText(user.email)).toBeInTheDocument();
    });

    it('TaskPanel signOut event is called', () => {
        // Arrange.
        // Act.
        render(
            <TaskPanel
                user={user}
                signOut={signOut}
                clearTaskList={clearTaskList}
                openTask={openTask}
            />,
        );

        screen.getByText('Sign Out').click();

        // Assert.
        expect(signOut).toHaveBeenCalledTimes(1);
        expect(clearTaskList).toHaveBeenCalledTimes(1);
    });

    it('TaskPanel openTask event is called', () => {
        // Arrange.
        // Act.
        render(
            <TaskPanel
                user={user}
                signOut={signOut}
                clearTaskList={clearTaskList}
                openTask={openTask}
            />,
        );

        screen.getByText('Create').click();

        // Assert.
        expect(openTask).toHaveBeenCalledTimes(1);
    });
});
