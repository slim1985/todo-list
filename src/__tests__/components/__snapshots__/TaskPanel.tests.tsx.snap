// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskPanel TaskPanel renders correctly with user 1`] = `
<DocumentFragment>
  <div
    class="flex justify-start sticky top-0 space-x-3 bg-white solid border-gray-400 border-b-2 retina:border-b-4"
  >
    <div
      class="flex w-2/4 justify-start items-center float-left"
    >
      <div
        class="ml-2 retina:ml-6"
      >
        <p
          class="font-semibold retina:text-5xl"
        >
          Test User
        </p>
        <p
          class="retina:mt-3 text-sm retina:text-5xl"
        >
          user@user.com
        </p>
        <button
          class="mt-2 retina:mt-5 retina:text-5xl"
        >
          Sign Out
        </button>
      </div>
    </div>
    <div
      class="flex w-2/4 justify-end"
    >
      <button
        class="size-20 retina:size-52 my-2 retina:my-4 mr-2 retina:mr-5 p-1 retina:text-5xl bg-gray-300 rounded-md retina:rounded-3xl"
      >
        Create
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TaskPanel TaskPanel renders correctly without user 1`] = `
<DocumentFragment>
  <div
    class="flex justify-start sticky top-0 space-x-3 bg-white solid border-gray-400 border-b-2 retina:border-b-4"
  />
</DocumentFragment>
`;
